strawberryfields/apps/sample.py

```diff
@@ -361,3 +377,42 @@ def vibronic(
     if np.any(t == 0):
         s = np.pad(s, ((0, 0), (0, n_modes))).tolist()
     return s
+
+
+def waw_matrix(A: np.ndarray, w: Union[np.ndarray, list]) -> np.ndarray:
+    r"""Rescale adjacency matrix to account for node weights.
+
+    Given a graph with adjacency matrix :math:`A` and a vector :math:`w` of weighted nodes,
+    this function rescales the adjacency matrix according to :cite:`banchi2019molecular`:
+
+    .. math::
+
+        A \rightarrow WAW,
+
+    with :math:`W=w_{i}\delta_{ij}` the diagonal matrix formed by the weighted nodes :math:`w_{i}`.
+    The rescaled adjacency matrix can be passed to :func:`sample`, resulting in a distribution that
+    increases the probability of observing a node proportionally to its weight.
+
+    **Example usage:**
+
+    >>> g = nx.erdos_renyi_graph(5, 0.7)
+    >>> a = nx.to_numpy_array(g)
+    >>> w = [10, 1, 0, 1, 1]
+    >>> a = waw_matrix(a, w)
+    >>> sample(a, 3, 4)
+    [[1, 1, 0, 1, 0], [1, 1, 0, 1, 0], [1, 1, 0, 0, 0], [1, 1, 0, 1, 0]]
+
+    Args:
+        A (array): adjacency matrix to rescale
+        w (array or list): vector of real node weights
+
+    Returns:
+        array: matrix rescaled according to the WAW encoding
+    """
+    if not np.allclose(A, A.T):
+        raise ValueError("Input must be a NumPy array corresponding to a symmetric matrix")
+
+    if isinstance(w, list):
+        w = np.array(w)
+
+    return (w * A).T * w
```

Comment:
Maybe I'm missing something (because the tests below pass), but is this doing WAW? I did a small test in NumPy:

```python
>>> import numpy as np
>>> A = np.random.random(size=[4,4])
>>> w = np.random.random(size=[4])
>>> (w * A).T * w == np.diag(w) @ A @ np.diag(w)
array([[ True, False, False, False],
       [False,  True, False, False],
       [False, False,  True, False],
       [False, False, False,  True]])
```

Reply:
It seems that the diagonal is rescaled correctly, but the off-diagonals differ. But let me know if I'm doing something silly 😆

Reply:
Your code block looks right, but in this case A is symmetric. We have:
`w * A =  A @ np.diag(w)`
`(w * A).T = np.diag(w) @ A.T`
`(w * A).T = np.diag(w) @ A`
`(w * A).T * w = np.diag(w) @ A.T @ np.diag(w)`
So the `(w * A).T * w` approach does require symmetry of `A`, and wouldn't work if `A` was complex and hermitian. However, I think we are reasonably restricting to graphs with real edge weights for now.

Although, perhaps going with `np.diag(w) @ A @ np.diag(w)` is simpler. One minor drawback of `np.diag(w) @ A @ np.diag(w)` is that it doesn't work vectors that are 2D arrays (e.g. row or column vector), while the `(w * A).T * w` approach does - but this setting should be relatively rare.